기본적인 SSE 샘플 코드 업로드

Bue-von-hon opened the pull request:
Motication:
한 명의 사용자가 여러 탭을 실행중인 상황을 가정.
이때 해당 사용자가 보고있는 게시글은 news1, news2 둘 중 하나라고 가정.
A라는 탭에서 news1을 B라는 탭에서 news2를 C라는 탭에서 news1을 보고있는 상황. A탭에서 news1의 조회수를 증가시키면 C탭에서는 조회수가 증가하지만, B에서는 증가하면 안됌. 그 반대의 경우도 마찬가지라 가정하고 코드를 작성.

Modification:
- application.yml 로깅 설정을 위해 추가 (info 레벨)
- build.gradle.kts 프로젝트를 위한 기본적인 의존성 추가
- index.html에 커넥션 요청과 조회수 증가를 위한 버튼 생성 - 커넥션 요청 시 쿼리 파라미터로 아티클 아이디와 탭 아이디를 전달 - 탭 아이디는 랜덤하게, 아티클 아이디는 news1 또는 news2로 생성 - 이벤트 리스너를 등록하여 콘솔 로그를 통해 조회수 카운트 확인 기능 구현
- IndexController: index.html 파일을 내려주기 위한 컨트롤러 구현
- SseController: 커넥션 요청과 조회수 증가 요청을 처리하기 위한 컨트롤러 구현
- SseEmitters: 각 아티클과 조회수가 매핑되는 해시맵, 탭과 매핑된 커넥션이 담긴 해시맵, 아티클이 열려있는 탭 목록을 담은 해시맵 구현 - add 메소드: 커넥션 요청이 오면 해시맵에 정보를 추가하는 기능 구현 - count 메소드: 조회수 증가 요청이 전달되면 해당 아티클이 열려있는 모든 커넥션에 조회수를 증가시키는 기능 구현

Result:
- 로깅 설정을 추가하여 로그 레벨이 info로 설정됨
- index.html에 커넥션 요청과 조회수 증가를 위한 버튼이 생성됨
- index.html에서 쿼리 파라미터를 통해 올바른 아티클 아이디와 랜덤하게 생성된 탭 아이디가 전달됨
- 이벤트 리스너를 통해 조회수가 증가하고, 콘솔 로그를 통해 확인할 수 있음
- 커넥션 요청 시 아티클 아이디, 탭 아이디 등을 조합해 커넥션을 유지
- 조회수 증가 요청시 아티클 아이디를 통해 해당 아티클이 열려있는 모든 탭에 조회수 증가 요청을 전송

Bue-von-hon committed:
Motication:
한 명의 사용자가 여러 탭을 실행중인 상황을 가정.
이때 해당 사용자가 보고있는 게시글은 news1, news2 둘 중 하나라고 가정.
A라는 탭에서 news1을 B라는 탭에서 news2를 C라는 탭에서 news1을 보고있는 상황.
A탭에서 news1의 조회수를 증가시키면 C탭에서는 조회수가 증가하지만, B에서는 증가하면 안됌.
그 반대의 경우도 마찬가지라 가정하고 코드를 작성.

Modification:
- application.yml 로깅 설정을 위해 추가 (info 레벨)
- build.gradle.kts 프로젝트를 위한 기본적인 의존성 추가
- index.html에 커넥션 요청과 조회수 증가를 위한 버튼 생성
    - 커넥션 요청 시 쿼리 파라미터로 아티클 아이디와 탭 아이디를 전달
    - 탭 아이디는 랜덤하게, 아티클 아이디는 news1 또는 news2로 생성
    - 이벤트 리스너를 등록하여 콘솔 로그를 통해 조회수 카운트 확인 기능 구현
- IndexController: index.html 파일을 내려주기 위한 컨트롤러 구현
- SseController: 커넥션 요청과 조회수 증가 요청을 처리하기 위한 컨트롤러 구현
- SseEmitters: 각 아티클과 조회수가 매핑되는 해시맵, 탭과 매핑된 커넥션이 담긴 해시맵,
    아티클이 열려있는 탭 목록을 담은 해시맵 구현
    - add 메소드: 커넥션 요청이 오면 해시맵에 정보를 추가하는 기능 구현
    - count 메소드: 조회수 증가 요청이 전달되면 해당 아티클이 열려있는 모든 커넥션에 조회수를 증가시키는 기능 구현

Result:
- 로깅 설정을 추가하여 로그 레벨이 info로 설정됨
- index.html에 커넥션 요청과 조회수 증가를 위한 버튼이 생성됨
- index.html에서 쿼리 파라미터를 통해 올바른 아티클 아이디와 랜덤하게 생성된 탭 아이디가 전달됨
- 이벤트 리스너를 통해 조회수가 증가하고, 콘솔 로그를 통해 확인할 수 있음
- 커넥션 요청 시 아티클 아이디, 탭 아이디 등을 조합해 커넥션을 유지
- 조회수 증가 요청시 아티클 아이디를 통해 해당 아티클이 열려있는 모든 탭에 조회수 증가 요청을 전송

ddaaac commented on src/main/java/com/example/youthcon/SseEmitters.java:
```diff
@@ -0,0 +1,61 @@
+    private final ConcurrentHashMap<String, List<String>> articleId2tabId = new ConcurrentHashMap<>();
+
+    SseEmitter add(final String tabId, final String articleId) {
+        final SseEmitter emitter = new SseEmitter();
```
여기서 타임아웃을 설정해줄 수 있을 것 같아요
- 어느 정도 타임아웃이 적당할지
- 기본 값은 어떤 값을 따르는지
- 타임아웃이 발생했을 때 어떤 동작이 발생하는지

등을 얘기해볼 수 있을 것 같습니다.

Bue-von-hon replied:
- 타임아웃 시간 설정
    - 5분이 어떨까요..? 유저의 사용성을 고려해서...
    - 너무 짧으면 재연결이 빈번하고, 너무 길면 유지해야 할 커넥션이 많아지는 트레이드 오프가...
- 타임아웃 기본값
    - MVC Java Config 또는 MVC 네임스페이스에 구성된 기본값이 사용 이것도 없으면 서버 기본값
    - 이렇게 SseEmitter의 생성자에 주석이 달려있었습니다.
- 타임아웃이 발생했을 때 동작 과정
    - 타임아웃시 실행시킬 콜백이 없으면 동작 안함
    - 콜백이 있으면 해당 콜백이 실행, 컨테이너 쓰레드에서 호출 (ResponseBodyEmitter#onTimeout 주석 참고)
    - 클라이언트에서는 타임아웃나면 재연결 요청을 보낸다고 하는데 이 부분은 잘 모르겠네요... [링크](https://tecoble.techcourse.co.kr/post/2022-10-11-server-sent-events/#:~:text=%ED%83%80%EC%9E%84%EC%95%84%EC%9B%83%EC%9D%B4%20%EB%B0%9C%EC%83%9D%ED%95%98%EB%A9%B4%20%EB%B8%8C%EB%9D%BC%EC%9A%B0%EC%A0%80%EC%97%90%EC%84%9C%20%EC%9E%AC%EC%97%B0%EA%B2%B0%20%EC%9A%94%EC%B2%AD%EC%9D%84%20%EB%B3%B4%EB%82%B4%EB%8A%94%EB%8D%B0%2C%20%EC%9D%B4%EB%95%8C%20%EC%83%88%EB%A1%9C%EC%9A%B4%20Emitter%20%EA%B0%9D%EC%B2%B4%EB%A5%BC%20%EB%8B%A4%EC%8B%9C%20%EC%83%9D%EC%84%B1%ED%95%98%EA%B8%B0%20%EB%95%8C%EB%AC%B8%EC%97%90(SseController%EC%9D%98%20connect()%EB%A9%94%EC%84%9C%EB%93%9C%20%EC%B0%B8%EC%A1%B0)%20%EA%B8%B0%EC%A1%B4%EC%9D%98%20Emitter%EB%A5%BC%20%EC%A0%9C%EA%B1%B0%ED%95%B4%EC%A3%BC%EC%96%B4%EC%95%BC%20%ED%95%A9%EB%8B%88%EB%8B%A4.)

ddaaac replied:
- 타임아웃이 일어나면 재연결 요청을 하게 됩니다. 사실상 관리해야 할 커넥션의 수는 동일하게 되겠죠. 타임아웃이 훨씬 길어도 될 것 같아요. 시연에서는 타임아웃을 짧게(1초 정도) 잡고 브라우저에서 1초마다 연결을 새로 하는 모습을 보여줄 수 있겠죠.
- 별도 처리가 없다면 부트 내장 톰캣의 기본값인 30을 따라간다고 하네요. `spring.mvc.async.request-timeout` 이 값으로 제어할 수 있다고 하구요. 개인적으로는 그냥 생성자에서 우리가 지정한 값을 명시해주는 게 유지보수할 때 좋다고 생각해요.
- 클라에서 강제로 타임아웃을 발생시키는 법은 그냥 서버를 내려버리는 방법이 있습니다. 이 때 클라에서 retry 간격은 어떻게 조절할 수 있을지도 찾아보시면 좋을 것 같아요.

아래는 서버를 내렸다가 다시 올렸을 때 요청의 성공/실패입니다. 시연 때 브라우저 개발자 콘솔의 네트워크 탭을 열어서 여기를 보여줘도 좋을 것 같아요.
<img width="193" alt="image" src="https://github.com/Bue-von-hon/SSE-samples/assets/43440195/e09312df-17ad-4d14-9a36-e3003c119120">

ddaaac commented on src/main/java/com/example/youthcon/SseEmitters.java:
```diff
@@ -0,0 +1,61 @@
+    SseEmitter add(final String tabId, final String articleId) {
+        final SseEmitter emitter = new SseEmitter();
+        articleId2tabId.computeIfAbsent(articleId, k -> new ArrayList<>()).add(tabId);
+        tabId2SseEmitter.putIfAbsent(tabId, emitter);
```
한 탭에서 기존 페이지를 연 상태에서 다른 페이지(다른 아티클)을 연 경우에 대한 처리도 필요할 것 같아요!

Bue-von-hon replied:
넵 해당 기능을 추가했습니다.

ddaaac replied:
방법에 대해서는 크게 고민해보지 않았는데 지금처럼 Article <-> Tab의 서로 참조하는 방식의 Map이 2개나 있는 건 유지보수 관점이나 가독성의 관점에서나 복잡하다는 생각이 들어요. 아래와 같이 콜백을 활용한 방식은 어떨까요?

```java
public class YourClass {
    private Map<String, Set<SseEmitter>> articleToEmitter = new HashMap<>();
    private Map<String, SseEmitter> tabToEmitter = new HashMap<>();

    public void add(String articleId, String tabId) {
        SseEmitter oldEmitter = tabToEmitter.get(tabId);
        if (oldEmitter != null) {
            oldEmitter.complete();
        }

        SseEmitter newEmitter = new SseEmitter();
        tabToEmitter.put(tabId, newEmitter);

        Set<SseEmitter> emitters = articleToEmitter.getOrDefault(articleId, new HashSet<>());
        emitters.add(newEmitter);
        articleToEmitter.put(articleId, emitters);

        cleanUpArticle(articleId, newEmitter);
    }

    public void cleanUpArticle(String articleId, SseEmitter emitter) {
        emitter.onCompletion(() -> {
            Set<SseEmitter> emitters = articleToEmitter.get(articleId);
            if (emitters != null) {
                emitters.remove(emitter);
            }
        });

        emitter.onError((error) -> {
            Set<SseEmitter> emitters = articleToEmitter.get(articleId);
            if (emitters != null) {
                emitters.remove(emitter);
            }
        });

        emitter.onTimeout(() -> {
            Set<SseEmitter> emitters = articleToEmitter.get(articleId);
            if (emitters != null) {
                emitters.remove(emitter);
            }
        });
    }
}

```

- 동기화 문제는 적절한 자료구조를 사용해서 해결해야할 것 같아요.
- SseEmitter의 HashCode를 예측하기 어렵기 때문에 SseEmitter를 감싼 객체를 하나 더 만들어서 어떤 식으로든 안정적인 HashCode를 제공해야할 것 같네요.
- 또는 적당한 일급 컬렉션을 활용해볼 수도 있을 것 같구요.
- 이 경우에 사용자가 tab을 닫으면 tabToEmitter에서 tab이 없어질 방법이 없기 때문에 tabToEmitter에 ttl을 걸어야 할 수도 있을 것 같습니다. 이 때 ttl 값은 앞서 코멘트 남긴 SseEmitter의 timeout 값과 같거나 살짝 크면 되겠죠. 어차피 timeout 값 뒤에 tab이 연결되어있다면 재연결 요청이 들어올테니깐요. (여기까진 다룰 필요가 있을까 싶기도 하구요. 실습은 안하고 말만 하고 넘어가도 될 것 같구요)

Bue-von-hon replied:
콜백을 이용하도록 수정했습니다.
동기화 문제는 conccruentHashMap을 이용했습니다.
안정적인  해시코드 문제는 잘 이해가 가지 않는데, 기존 방식에 어떤 문제가 있는걸까요..?
TTL은 내일 추가해보겠습니다...ㅎ

현재는 타임아웃 설정 시간이 지나면, 유저가 재 요청을 보내고,
재요청시 기존의 커넥션을 날려버립니다(complete 호출)
클라이언트쪽에서는 서버에서 커넥션을 정리하면 onerror 메소드가 호출되는데요
이때 sse 커넥션 닫아버리게 코드를 추가했습니다.
결과적으로 타임아웃 시 에러가 출력되는데, 이 오류는 TTL 설정으로 에러가 해결될 것 같습니다

ddaaac commented on src/main/java/com/example/youthcon/SseEmitters.java:
```diff
@@ -0,0 +1,61 @@
+
+        final List<String> tabIdList = articleId2tabId.get(articleId);
+        tabIdList.forEach(tabId -> {
+            final SseEmitter emitter = tabId2SseEmitter.get(tabId);
```
null 처리에 신경써볼 수 있을 것 같습니다.

Bue-von-hon replied:
null 처리 로직 추가했습니다!

ddaaac commented on src/main/java/com/example/youthcon/SseEmitters.java:
```diff
@@ -0,0 +1,61 @@
+        tabIdList.forEach(tabId -> {
+            final SseEmitter emitter = tabId2SseEmitter.get(tabId);
+            try {
+                emitter.send(SseEmitter.event()
```
이 부분은 궁금한 점인데 만약 사용자가 탭을 닫았고, 그 탭에 send를 하게 될 경우 서버에서는 탭을 닫았다는 사실을 감지할 수 있나요?

Bue-von-hon replied:
탭을 닫은건 감지 못합니다만,
탭이 닫힌 이후 아무런 응답이 없어 타임아웃이 호출되는것 같아요.
타임아웃이 호출된 경우 탭이 닫혀있는면, 자원을 회수합니다.
타임아웃이 호출된 경우 탭이 열려 있으면, 기존 커넥션이 그냥 유지되고 있어요.

ddaaac commented on src/main/resources/templates/index.html:
```diff
@@ -0,0 +1,69 @@
+
+  document.getElementById('connectButton').addEventListener('click', handleConnect);
+  document.getElementById('countButton').addEventListener('click', handleCountClick);
+</script>
```
브라우저를 닫을 때 연결을 끊는 api와 프런트 코드도 있으면 좋을 것 같아요.
서버에서 끊어짐을 감지할 방법이 있다면 더 좋겠지만 제가 알기로는 클라에서 연결 끊김을 감지하기 쉽지 않은 것으로 알고 있습니다(이 부분도 확인이 필요하겠네요).

Bue-von-hon replied:
일단 window.onbeforeunload 이 메소드로 연결 끊김 감지 기능 구현했습니다...
브라우저 닫힐때, 새로고칠 할 때 cleanup 되네요
닫힐때만 하는건 잘 모르겠습니다..😅

ddaaac replied:
onbeforeunload 이벤트 잘 찾아주셨네요 👍 
위에 코멘트 남긴대로 구현하면 요 로직은 필요없게 될 것 같아요. 자원을 얼마나 최적화할거냐에 대한 얘기를 해볼 수 있을 것 같은데, 핸즈온 할 때는 굳이 보여주지 않아도 되는 부분 같습니다!

ddaaac reviewed (commented):
안녕하세요~

빠르게 뚝딱 잘 만들어주셨네요! 
연결 끊김 상황에 대해서 고려해보면 좋을 것 같아서 몇 가지 코멘트 남겼으니 확인해주세요~

Bue-von-hon reviewed (commented):
@ddaaac 남겨주신 피드백을 해결했습니다 PTAL 🙏

Bue-von-hon committed:
Fixed based on code review

ddaaac reviewed (commented):
저장 구조에 대해서 조금 고민을 해봤습니다.
혹시 로직에 구멍이 있을 수 있으니 천천히 확인해주세요~

Bue-von-hon committed:
Fixed based on code review

Modification:
- index.html: 아티클 변경 시, (old, new) 아티클 아이디를 서버로 보내 커넥션을 변경하기 쉽게 수정, onerror 콜백을 이용해 타임아웃 시 처리를 할 수 있도록 추가
- SseController.java: 새로운/기존 아티클 아이디 2개를 받도록 수정
- SseEmitters.java: 콜백을 이용하여 커넥션을 처리하도록 변경, 콜백 세팅해주는 메소드 추가, 아티클 변경 처리를 old/new 아티클 아이디를 받아서 로직 간소화

Bue-von-hon committed:
불필요한 코드 삭제

Bue-von-hon committed:
타임아웃 시 커넥션 닫게 수정

Bue-von-hon commented:
@ddaaac 말씀해주신 부분들을 추가했고, 사소한 질문을 남겼습니다 🙏

Bue-von-hon committed:
Modification:
- 재요청 시간 설정
- 타임 아웃 시간 수정
- 캐시를 이용해 TTL 설정

Result:
재요청, 타임아웃, TTL을 동일하게 설정했습니다.

Bue-von-hon commented:
흠.... 캐시를 이용해 TTL을 구현했습니다.
타임아웃, 재요청, TTL 모두를 동일하게 설정했는데도 불구하고...
빠르게 조회수 증가 요청을 전송하면, 이미 리스폰스 바디가 완료되었다는 에러가 출력되네요(타임아웃과 캐시를 동일하게 설정하면 발생하네요)
```java
ResponseBodyEmitter has already completed
	at org.springframework.util.Assert.state(Assert.java:97) ~[spring-core-6.0.10.jar:6.0.10]
	at org.springframework.web.servlet.mvc.method.annotation.ResponseBodyEmitter.send(ResponseBodyEmitter.java:196) ~[spring-webmvc-6.0.10.jar:6.0.10]
	at org.springframework.web.servlet.mvc.method.annotation.SseEmitter.send(SseEmitter.java:127) ~[spring-webmvc-6.0.10.jar:6.0.10]
	at com.example.youthcon.SseEmitters.lambda$count$4(SseEmitters.java:104) ~[main/:na]
	at java.base/java.lang.Iterable.forEach(Iterable.java:75) ~[na:na]
	at com.example.youthcon.SseEmitters.count(SseEmitters.java:98) ~[main/:na]
	at com.example.youthcon.SseController.count(SseController.java:44) ~[main/:na]
```

혹은 아래 코드의 complete를 호출하는 부분에서 The response object has been recycled and is no longer associated with this facade 에러가 출력되네요...(캐시 만료를 타임아웃보다 길게 잡은 케이스에서 자주 발생합니다)
```java
public SseEmitter connect(final String tabId, final String articleId) {
        final SseEmitter oldEmitter = tabId2SseEmitter.getIfPresent(tabId);

        if (oldEmitter != null) {
            oldEmitter.complete();
        }
```

아무래도 콜백으로 변경되면서, 동시성 이슈가 발생한 것 같습니다...
콜백쪽에는 동시성 제어 로직을 추가했는데... SseEmitters 클래스의 내부 변수 접근 로직이 의심스럽네요...
이건 내일 마저 해보겠습니다...

ddaaac commented on src/main/java/com/example/youthcon/SseEmitters.java:
```diff
@@ -0,0 +1,147 @@
+    // todo(hun): 아티클별 조회수
+    private static final Map<String, AtomicLong> articleId2AtomicLong = new ConcurrentHashMap<>();
+
+    private final LoadingCache<String, SseEmitter> tabId2SseEmitter = CacheBuilder.newBuilder()
+                                                                    .expireAfterAccess(10000L, TimeUnit.MILLISECONDS)
+                                                                    .build(
+                                                                            new CacheLoader<String, SseEmitter>() {
+                                                                                @Override
+                                                                                public SseEmitter load(String key) {
+                                                                                    // 캐시 항목이 없거나 만료된 경우에 값을 로드하는 코드
+                                                                                    return null;
+                                                                                }
+                                                                            }
```
```suggestion
    private final Cache<String, SseEmitter> tabId2SseEmitter = CacheBuilder.newBuilder()
                                                                    .expireAfterAccess(10000L, TimeUnit.MILLISECONDS)
                                                                    .build(
```

CacheLoader의 load 설명을 보면 null을 리턴하지 말라고 되어있습니다. 이 기능이 기본값을 넣어주는 기능이기 때문에 기본값을 넣지 않을 거라면 사용하지 않아도 될 것 같아요.

Bue-von-hon replied:
넵 수정했습니다!

ddaaac commented on src/main/java/com/example/youthcon/SseController.java:
```diff
@@ -0,0 +1,61 @@
+        log.info("tabId : {}", tabId);
+        final SseEmitter emitter = sseEmitters.connect(tabId, articleId);
+        try {
+            final SseEmitter.SseEventBuilder eventBuilder = SseEmitter.event()
+                                                                      .name("connect")
+                                                                      .data("connected!")
+                                                                      .reconnectTime(10000L);
```
> 빠르게 조회수 증가 요청을 전송하면, 이미 리스폰스 바디가 완료되었다는 에러가 출력되네요

<img width="801" alt="image" src="https://github.com/Bue-von-hon/SSE-samples/assets/43440195/3aad1c4b-149c-4c92-8359-2f526e508592">

빠르게와 상관없이 reconnection 값을 길게 잡아서 생기는 이슈일 것 같아요. 저 구간들의 길이를 모두 서버에서 제어해주고 있는데요, 어떤 값이 어떤 영향을 미치는지 확인해보시면 좋을 것 같습니다.

Bue-von-hon replied:
서버-클라이언트 연결이 되어있는 상태에서, 서버를 죽이면 클라이언트가 설정된 시간 텀마다 리트라이하는걸 확인했습니다.
처음 연결 이후 타임아웃 설정된 시간 마다 연결 요청, 에러 상황에서(서버를 끄거나..) 리트라이 시간 텀마다 요청을 보내는 식으로 동작하는 것 같아요. 생각보다 복잡하네요..ㅎㅎ

ddaaac commented on src/main/java/com/example/youthcon/SseEmitters.java:
```diff
@@ -0,0 +1,147 @@
+    private final Map<String, Set<SseEmitter>> articleToEmitter = new ConcurrentHashMap<>();
+
+    public SseEmitter connect(final String tabId, final String articleId) {
+        final SseEmitter oldEmitter = tabId2SseEmitter.getIfPresent(tabId);
+
+        if (oldEmitter != null) {
+            oldEmitter.complete();
+        }
```
> 혹은 아래 코드의 complete를 호출하는 부분에서 The response object has been recycled and is no longer associated with this facade 에러가 출력되네요...(캐시 만료를 타임아웃보다 길게 잡은 케이스에서 자주 발생합니다)

요건 자연스러운 현상같아요. 
<img width="193" alt="image" src="https://github.com/Bue-von-hon/SSE-samples/assets/43440195/cae64a39-5ada-421a-80fc-fb7717d72326">
서버에서 타임아웃을 10000ms로 잡아도 실제로 네트워크 탭을 확인하면 10초보다 조금 길게 커넥션이 유지됩니다. 
<img width="846" alt="image" src="https://github.com/Bue-von-hon/SSE-samples/assets/43440195/4d8ddae4-909e-4a5a-a98a-47d42f951ab2">
만약 저 차이나는 구간에 새로운 커넥션이 맺어지면 compelete이 동작하지 못해서 push가 2번 가는 현상이 일어날거에요. 이 어플리케이션에서는 크리티컬하지 않기 때문에 살짝 길게 잡는 것과 동일하게 잡는 것이 크게 차이는 없을 것 같네요.

여튼, 다시 주신 질문으로 돌아가면 tab - emitter를 보관하는 이유는 새로운 emitter를 만들 때 기존 emitter를 만료시켜서 여러번 push가 가는 걸 방지하기 위함입니다. timeout은 시간 단위로 길게 잡기 때문에 새로운 emitter를 만료시키지 않는다면 너무 많은 emitter들을 관리해야하거든요. 
안전하게 기존 emitter를 만료시키기 위해 ttl을 timeout보다 크게 설정했다면 당연히 만료된 emitter를 보관하는 경우도 있을텐데요, 이를 다시 만료시키려할 때 저런 에러가 난다면 이는 충분히 발생할 수 있는 케이스니 그냥 무시해도 될 것 같아요.

Bue-von-hon replied:
실제로 타임아웃 로그를 확인하면서 실습해보니 문제가 없네요..!

ddaaac commented on src/main/java/com/example/youthcon/SseEmitters.java:
```diff
@@ -0,0 +1,147 @@
+        });
+    }
+
+    public void count(final String articleId) {
+        final AtomicLong counter = articleId2AtomicLong.computeIfAbsent(articleId, k -> new AtomicLong());
+        final long count = counter.incrementAndGet();
```
핸즈온에서 발표할 때는 여기서 값을 늘리는 게 아닌 DB에서 좋아요 수를 꺼내와서 +1하고 이를 브라우저에 뿌린다는 설명 정도를 할 수 있겠네요.

ddaaac commented on src/main/java/com/example/youthcon/SseEmitters.java:
```diff
@@ -0,0 +1,147 @@
+                                                                            }
+                                                                    );
+
+    private final Map<String, Set<SseEmitter>> articleToEmitter = new ConcurrentHashMap<>();
```
[HashSet](https://tecoble.techcourse.co.kr/post/2020-07-29-equals-and-hashCode/)은 hashCode를 활용하여 동작합니다. 따라서 SseEmitter의 hashCode가 올바르게 정의되어있는지 확인해봐야돼요.

Bue-von-hon replied:
커넥션 클래스로 감싸고 equals, hashcode 재정의하는 방향으로... 수정했습니다!

ddaaac commented on src/main/java/com/example/youthcon/SseEmitters.java:
```diff
@@ -0,0 +1,147 @@
+        setCallback(articleId, newEmitter);
+        tabId2SseEmitter.put(tabId, newEmitter);
+
+        final Set<SseEmitter> connections = articleToEmitter.getOrDefault(articleId, new HashSet<>());
+        connections.add(newEmitter);
+        articleToEmitter.put(articleId, connections);
```
아래에서 동기화 처리를 잘해주셨어요. 여기도 필요할 것 같습니다.

Bue-von-hon replied:
추가했습니다!

ddaaac reviewed (commented):
대충 뼈대는 거의 다 나온 느낌이네요!

이번에 나온 이슈만 마무리하면 코드 정리 + 발표 준비를 할 수 있을 것 같아요.

화이팅하세요 💪

Bue-von-hon committed:
Modification:
- index.html: 현재 탭에서 아티클을 변경한 경우, 새로운 커넥션 요청을 보내게 변경
- Connection.java: emitter를 쓰레드-세이프하게 다루기 위한 클래스, tab id를 이용해 동등성 및 동일성 보장
- SseController.java: 불필요한 메소드 삭제
- SseEmitters.java: 명시적 락을 활용해 공유변수에 접근하도록 수정, emitter를 직접 사용하는 대신 connection을 사용하게 수정, 공통으로 사용하는 로직 메소드로 추출, 캐시가 만료되어도 그대로 남아있는 문제를 명시적으로 cleanup하여 해결

Bue-von-hon commented:
@ddaaac 문제가 있던 부분을 모두 고쳤고, 전반적으로 읽을만하게 코드를 수정했습니다!ㅋㅋ
2차 리허설에서 피드백을 받은게 있어서... 조회수 외에 다른걸로도 해볼지 고민을 해야....

Bue-von-hon committed:
간단한 Sse 테스트 코드와 이를 위한 의존성 추가

Commit:
Fix dependency and minor things

Modification:
- Fix(build.gradle.kts): 자바 17 관련 의존성 수정 및 버전업
- Fix(Connection, SseEmitters): 이미터를 직접 주입하지 않고, 타임아웃만 설정하게 수정
